halo2-ecc/src/bn254/pairing.rs

```diff
@@ -521,4 +521,29 @@ impl<'chip, F: PrimeField> PairingChip<'chip, F> {
         // final_exp implemented in final_exp module
         fp12_chip.final_exp(ctx, f0)
     }
+
+    /*
+     * Conducts an efficient pairing check e(P, Q) = e(S, T) using only one
+     * final exponentiation. In particular, this constraints
+     * (e'(-P, Q)e'(S, T))^x = 1, where e' is the optimal ate pairing without
+     * the final exponentiation. Reduces number of necessary advice cells by
+     * ~30%.
+     */
+    pub fn pairing_check(
+        &self,
+        ctx: &mut Context<F>,
+        Q: &EcPoint<F, FqPoint<F>>,
+        P: &EcPoint<F, FpPoint<F>>,
+        T: &EcPoint<F, FqPoint<F>>,
+        S: &EcPoint<F, FpPoint<F>>,
+    ) {
+        let ecc_chip_fp = EccChip::new(self.fp_chip);
+        let negated_P = ecc_chip_fp.negate(ctx, P);
+        let mml = self.multi_miller_loop(ctx, vec![(&negated_P, Q), (S, T)]);
+        let fp12_chip = Fp12Chip::<F>::new(self.fp_chip);
+        let fe = fp12_chip.final_exp(ctx, mml);
+        let fv_chip = FieldVectorChip::new(self.fp_chip);
```

Comment:
Can we re-use `fp12_chip` to do the `assign_constant` without `fv_chip`?

Reply:
Done! Apologies for the delay. Thank you for reviewing :)